### ISSUE TITLE
Performance issue with 3k+ rows of table

### ISSUE THREAD

**Issue by iman38** [gorules/zen#276]
Hi @stefan-gorules ,

I'm curious about the performance of evaluating a rule.

Attached is a sample of the json file I used: [commission_rule.json](https://github.com/user-attachments/files/17868867/commission_rule.json)

Using this param (in the online editor):
```json
{
  "vendorId": 3,
  "route": "GTOKDI",
  "subclass": "WY"
}
```
it took +- 2.575201ms

while 
```json
{
  "vendorId": 1,
  "pcc": "7KOD",
  "origin": {
    "airport": "CGK"
  },
  "carrier": {
    "marketing": "AC"
  }
}
```
only took +- 141.436µs

Is this the type of performance I should expect or do I miss something to optimize for best performance? Evaluating on hundreds of params took quite a hit on overall performance.

Also, is there any dev friendly changelog of releases? 
I'm currently using version 0.29.1 (since 0.30 or later breaking my code, e.g. on 0.29.1 .evaluate accepting param from json! macro, 
while I see it's now a new Variable type).

Thank you for such a great crate.

Best,
Iman

**Comment by stefan-gorules** [comment]
Hi @iman38,

Thanks for sharing. It's quite difficult to get better performance in this case as we are doing live interpretation of the values on the go. To give you some insight

```
Columns: 13

First case index: 1418
Second case index: 8

First case cells: 18,434
Second case cell: 104
```

Not all cells will be evaluated as a lot of them are empty, but generous guess is that around 3,000 expressions are evaluated in first case, which gets evaluated in under 3ms. This includes every cell needing to do lexing, parsing, compiling and interpretation on the fly.

We could potentially explore some performance optimisations for your case, but as this isn't on our roadmap nor an immediate priority, it could be explored for prioritisation on commercial basis (in which case we could set something up if you reach out to us via hi@gorules.io). I assume we could get x3 to x4 the performance by caching expression results in your case.

As for the dev friendly changelog, unfortunately we don't have that yet, but we will try to prioritise it during 2025.

For Variable struct, you can simply convert from serde_json::Value to Variable using `From` trait.

You are welcome!

**Comment by iman38** [comment]
Hi @stefan-gorules 

Thanks for your response. sorry for the late response. I did end up with caching the result. To be fair, the realtime execution tooks around 50-200ms, but I was looking to get <10ms response from our API.

Interesting that you have commercial offer. will definitely explore this option when we need the support. Thanks for sharing.

I tried upgrading to 0.34.1, but found that I not only need to convert the param to Variables. I need adjusting the .evaluate calling as I'm on async multithreaded app (and not sure what else). 0.29 works well for my case, so I guess I will stay on that version till I have some time to spare for updating the "breaking" crates updates
